### PR TITLE
Add some basic tests for SymEngine

### PR DESCRIPTION
**Pull request by jppelteret**
Blocked by #7810

### CODE REVIEW

**Comment by jppelteret** [comment]
@luca-heltai FYI

**Comment by luca-heltai** [comment]
Thanks @jppelteret ! This is a lot of work. I'm looking forward to playing around with this... 

:)

**Comment by jppelteret** [comment]
Just wait till you see the rest of it ;-)

Within the next few days I'll lay out a plan for the PR's within the project
https://github.com/dealii/dealii/projects/14
so that you can see where I'm going with this.

**Comment by jppelteret** [comment]
@tamiko I needed to amend the CMake module for SymEngine. The `SYMENGINE_LIBRARIES` variable was only populated with the string `symengine`, rather than the library path itself. This is now fixed but using our own macro to get the full library path.

File: cmake/modules/FindSYMENGINE.cmake [hunk 1]
```diff
@@ -118,7 +118,19 @@ STRING(REGEX REPLACE
   )
 REMOVE_DUPLICATES(_symengine_include_dirs)
 
-SET(_symengine_libraries ${SYMENGINE_LIBRARIES})
+#
+# The SYMENGINE_LIBRARIES variable configured by SymEngine only lists the
+# libraries, but does not set their paths. So we configure this outselves.
```
Comment (isuruf) [inline cmake/modules/FindSYMENGINE.cmake]: Are you sure about this? `SYMENGINE_LIBRARIES` is supposed to include a mix of targets (with the location set in the target) and absolute paths.

Comment (jppelteret) [reply cmake/modules/FindSYMENGINE.cmake]: Yeah, I am sure:
```
#
# Include the SymEngine:
#
FIND_PACKAGE(SymEngine ${SYMENGINE_MINIMUM_REQUIRED_VERSION}
  CONFIG QUIET
  HINTS ${SYMENGINE_DIR}
  PATH_SUFFIXES lib/cmake/symengine
  NO_SYSTEM_ENVIRONMENT_PATH
  )
  
MESSAGE(STATUS "LISTING THE SYMENGINE_LIBRARIES: ${SYMENGINE_LIBRARIES}")
```
prints
```
-- LISTING THE SYMENGINE_LIBRARIES: symengine
```
with https://github.com/symengine/symengine/commit/e66c30c60da65d4a0a058fd992c3c2f6b20be3d6

But I understand that this is probably not what is intended to happen on this line:
https://github.com/symengine/symengine/blob/master/cmake/SymEngineConfig.cmake.in#L101
Missing `" "` maybe?

Comment (jppelteret) [reply cmake/modules/FindSYMENGINE.cmake]: [Here's the `SymEngineConfig.cmake`](https://gist.github.com/jppelteret/eb31b8de6c8410a051394add4fb9110f) that's generated for me from  symengine/symengine@e66c30c. The following lines are suspiciously empty:
```
set(SYMENGINE_LLVM_LIBRARIES )
set(SYMENGINE_LLVM_COMPONENTS )

if (NOT "<non-empty boost path>" STREQUAL "")
    ...
    set(SYMENGINE_LIBRARIES ${SYMENGINE_LIBRARIES} )
endif()

```

Comment (isuruf) [reply cmake/modules/FindSYMENGINE.cmake]: Try doing,
```
foreach (LIB ${SYMENGINE_LIBRARIES})
    if ("${LIB}" STREQUAL symengine)
        get_property(SYMENGINE_LOC TARGET ${LIB} PROPERTY LOCATION)
        message("${SYMENGINE_LOC}")
    endif()
endforeach()
```

and you'll see that `symengine` is a target even though it prints as `symengine`

File: cmake/modules/FindSYMENGINE.cmake [hunk 2]
```diff
@@ -118,7 +117,22 @@ STRING(REGEX REPLACE
...
 REMOVE_DUPLICATES(_symengine_include_dirs)
 
-SET(_symengine_libraries ${SYMENGINE_LIBRARIES})
+#
+# Get the full path for the SYMENGINE_LIBRARIES. We could also do this using
+# DEAL_II_FIND_LIBRARY, but since these libraries are CMake targets we can
+# query them directly for this information.
+# 
+FOREACH(SYMENGINE_LIBRARY_NAME ${SYMENGINE_LIBRARIES})
+  GET_PROPERTY(SYMENGINE_LIBRARY TARGET ${SYMENGINE_LIBRARY_NAME} PROPERTY LOCATION)
```
Comment (isuruf) [inline cmake/modules/FindSYMENGINE.cmake]:
```suggestion
   if (TARGET ${SYMENGINE_LIBRARY_NAME})
       GET_PROPERTY(SYMENGINE_LIBRARY TARGET ${SYMENGINE_LIBRARY_NAME} PROPERTY LOCATION)
   else ()
       SET(SYMENGINE_LIBRARY ${SYMENGINE_LIBRARY_NAME})
   endif()
```

Comment (jppelteret) [reply cmake/modules/FindSYMENGINE.cmake]: Thanks. I don't see any entries other than `symengine` appear in `SYMENGINE_LIBRARIES`, so its been hard to sanity check this.

Comment (isuruf) [reply cmake/modules/FindSYMENGINE.cmake]: You are probably using a SymEngine build with no library dependencies (except LLVM).

Comment (jppelteret) [reply cmake/modules/FindSYMENGINE.cmake]: Yeah, that must be it. I use boostmp for integers, LLVM, and nothing else. I'll check with GMP etc. at a later stage. Thanks for the tip :-)

Comment (jppelteret) [reply cmake/modules/FindSYMENGINE.cmake]: BTW. I wrote all of the CMake stuff about 2 years ago, so a lot has changed since then!

**Comment by jppelteret** [comment]
@luca-heltai Are you happy with these tests?

**Comment by jppelteret** [comment]
Can we please merge this?

**Pull request merged**